src/analysis/lattice.h

```diff
@@ -19,37 +20,89 @@ constexpr bool has_getLeastUpperBound = std::
   is_invocable_r<void, decltype(&T::getLeastUpperBound), T, const T&>::value;
 template<typename T>
 constexpr bool has_isTop =
-  std::is_invocable_r<bool, decltype(T::isTop), const T&>::value;
+  std::is_invocable_r<bool, decltype(&T::isTop), T>::value;
 template<typename T>
 constexpr bool has_isBottom =
-  std::is_invocable_r<bool, decltype(T::isBottom), const T&>::value;
+  std::is_invocable_r<bool, decltype(&T::isBottom), T>::value;
 
 template<typename T>
 constexpr bool is_lattice =
   has_compare<T>&& has_getLeastUpperBound<T>&& has_isTop<T>&& has_isBottom<T>;
 
-// Represents a powerset lattice element (i.e. a set) as a bitvector. A true
-// means that an element is present in the set.
-template<size_t N> struct BitsetPowersetLattice {
-  std::bitset<N> value;
+// Represents a powerset lattice which is constructed from a finite set which
+// can be represented by a bitvector. Set elements are represented by
+// FinitePowersetLattice::Element, which represents members present in each
+// element by bits in the bitvector.
+class FinitePowersetLattice {
+
+  // The size of the set that the powerset lattice was created from. This is
+  // equivalent to the size of the Top lattice element.
+  size_t setSize;
+
+public:
+  FinitePowersetLattice(size_t setSize) : setSize(setSize) {}
+
+  // This represents an element of a powerset lattice. The element is itself a
+  // set which has set members. The bitvector tracks which possible members of
+  // the element are actually present.
+  struct Element {
```

Comment:
This could now be
```suggestion
  struct Element : public std::vector<bool> {
```
That would avoid needing to declare `count` etc as we can use them from the parent.

Reply:
Ah, good points. sgtm.